In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
data=pd.read_csv('data - data.csv')

In [4]:
print(data.shape)

(43431, 19)


In [5]:
print(data.keys())

Index(['id', 'user_id', 'vehicle_model_id', 'package_id', 'travel_type_id',
       'from_area_id', 'to_area_id', 'from_city_id', 'to_city_id', 'from_date',
       'to_date', 'online_booking', 'mobile_site_booking', 'booking_created',
       'from_lat', 'from_long', 'to_lat', 'to_long', 'Car_Cancellation'],
      dtype='object')


In [6]:
print(data.vehicle_model_id)

0        28
1        12
2        12
3        12
4        12
5        12
6        12
7        12
8        65
9        12
10       12
11       12
12       12
13       12
14       28
15       12
16       12
17       24
18       12
19       12
20       87
21       28
22       12
23       12
24       54
25       12
26       12
27       12
28       28
29       12
         ..
43401    65
43402    89
43403    65
43404    12
43405    12
43406    12
43407    87
43408    12
43409    12
43410    12
43411    65
43412    12
43413    12
43414    12
43415    12
43416    12
43417    12
43418    12
43419    12
43420    65
43421    12
43422    12
43423    24
43424    12
43425    12
43426    12
43427    12
43428    12
43429    87
43430    12
Name: vehicle_model_id, Length: 43431, dtype: int64


In [7]:
print(data.Car_Cancellation)

0        0
1        0
2        0
3        0
4        0
5        0
6        0
7        0
8        0
9        0
10       0
11       0
12       0
13       0
14       0
15       0
16       0
17       0
18       0
19       0
20       0
21       1
22       0
23       0
24       0
25       0
26       0
27       0
28       0
29       0
        ..
43401    0
43402    0
43403    0
43404    0
43405    0
43406    1
43407    0
43408    1
43409    0
43410    0
43411    1
43412    0
43413    0
43414    1
43415    0
43416    0
43417    0
43418    0
43419    0
43420    0
43421    0
43422    1
43423    0
43424    0
43425    0
43426    0
43427    0
43428    0
43429    0
43430    0
Name: Car_Cancellation, Length: 43431, dtype: int64


Bookings-> online booking- 0(no) 1(yes)......number of each
        ->mobile site booking- 0(no) 1(yes)......number of each
from_date- date and time of (pickup)
to_date- date and time of (drop)
booking_created- date amd time of booking
from_lat to_lat from_long to_long- from and to of lat and long
Car_cancellation- 0(not cancelled) 1(cancelled)....number of cancellations-varying with location(pickup...i guess), type of booking, time



number of cancellations- hurry- more demand

In [8]:
data

id  user_id  vehicle_model_id  package_id  travel_type_id  \
0      132512    22177                28         NaN               2   
1      132513    21413                12         NaN               2   
2      132514    22178                12         NaN               2   
3      132515    13034                12         NaN               2   
4      132517    22180                12         NaN               2   
5      132518    17712                12         NaN               2   
6      132519    22172                12         NaN               1   
7      132520    22181                12         NaN               2   
8      132521    22182                65         2.0               3   
9      132522    22184                12         NaN               2   
10     132523     4941                12         NaN               2   
11     132524    17037                12         NaN               2   
12     132525    22185                12         NaN               2   
13     132526    22186                12         NaN               2   
14     132527       16                28         NaN               2   
15     132528    15097                12         NaN               2   
16     132530      761                12         NaN               2   
17     132531    22189                24         NaN               1   
18     132533    17226                12         NaN               2   
19     132534      868                12         NaN               2   
20     132535    22190                87         2.0               3   
21     132536       16                28         NaN               2   
22     132537    21558                12         NaN               2   
23     132538    21995                12         1.0               3   
24     132539    22192                54         NaN               1   
25     132540    22193                12         NaN               2   
26     132541    20430                12         NaN               2   
27     132542    22194                12         NaN               2   
28     132543    21716                28         NaN               2   
29     132544    38966                12         NaN               2   
...       ...      ...               ...         ...             ...   
43401  185909    27899                65         7.0               3   
43402  185910    47996                89         1.0               3   
43403  185911    41939                65         7.0               3   
43404  185912    48711                12         NaN               2   
43405  185914    41939                12         4.0               3   
43406  185915    40643                12         3.0               3   
43407  185916    48714                87         NaN               1   
43408  185917    40643                12         3.0               3   
43409  185918    41939                12         4.0               3   
43410  185919    41939                12         4.0               3   
43411  185920    31846                65         4.0               3   
43412  185921    44027                12         NaN               2   
43413  185922    48718                12         NaN               2   
43414  185923    34127                12         NaN               2   
43415  185924    48719                12         1.0               3   
43416  185926    36929                12         NaN               2   
43417  185927    41939                12         4.0               3   
43418  185928    48721                12         NaN               2   
43419  185929    46807                12         NaN               2   
43420  185930    36265                65         2.0               3   
43421  185931    48723                12         NaN               2   
43422  185932    48724                12         NaN               2   
43423  185933    48725                24         NaN               2   
43424  185935    48726                12         2.

In [18]:
data.isnull().sum(axis=0)

id                         0
user_id                    0
vehicle_model_id           0
package_id             35881
travel_type_id             0
from_area_id              88
to_area_id              9138
from_city_id           27086
to_city_id             41843
from_date                  0
to_date                17890
online_booking             0
mobile_site_booking        0
booking_created            0
from_lat                  93
from_long                 93
to_lat                  9138
to_long                 9138
Car_Cancellation           0
dtype: int64

Of 43,431 entries in dataset,package_id has maximum NaN. Better to drop column. 

In [31]:
mod_data=data.drop(['package_id'],axis=1)

In [17]:
mod_data

id  user_id  vehicle_model_id  travel_type_id  from_area_id  \
0      132512    22177                28               2          83.0   
1      132513    21413                12               2        1010.0   
2      132514    22178                12               2        1301.0   
3      132515    13034                12               2         768.0   
4      132517    22180                12               2        1365.0   
5      132518    17712                12               2        1021.0   
6      132519    22172                12               1         571.0   
7      132520    22181                12               2        1192.0   
8      132521    22182                65               3         448.0   
9      132522    22184                12               2         516.0   
10     132523     4941                12               2         150.0   
11     132524    17037                12               2         455.0   
12     132525    22185                12               2        1166.0   
13     132526    22186                12               2         793.0   
14     132527       16                28               2        1063.0   
15     132528    15097                12               2        1102.0   
16     132530      761                12               2         814.0   
17     132531    22189                24               1        1383.0   
18     132533    17226                12               2        1353.0   
19     132534      868                12               2         297.0   
20     132535    22190                87               3         471.0   
21     132536       16                28               2          58.0   
22     132537    21558                12               2         540.0   
23     132538    21995                12               3        1286.0   
24     132539    22192                54               1         515.0   
25     132540    22193                12               2         142.0   
26     132541    20430                12               2         393.0   
27     132542    22194                12               2          89.0   
28     132543    21716                28               2        1237.0   
29     132544    38966                12               2          61.0   
...       ...      ...               ...             ...           ...   
43401  185909    27899                65               3        1038.0   
43402  185910    47996                89               3         328.0   
43403  185911    41939                65               3        1038.0   
43404  185912    48711                12               2         238.0   
43405  185914    41939                12               3        1281.0   
43406  185915    40643                12               3         136.0   
43407  185916    48714                87               1        1180.0   
43408  185917    40643                12               3         136.0   
43409  185918    41939                12               3        1281.0   
43410  185919    41939                12               3        1281.0   
43411  185920    31846                65               3         541.0   
43412  185921    44027                12               2        1384.0   
43413  185922    48718                12               2        1377.0   
43414  185923    34127                12               2         376.0   
43415  185924    48719                12               3         768.0   
43416  185926    36929                12               2         396.0   
43417  185927    41939                12               3        1390.0   
43418  185928    48721                12               2        1263.0   
43419  185929    46807                12               2         571.0   
43420  185930    36265                65               3        1237.0   
43421  185931    48723                12               2        1371.0   
43422  185932    48724                12               2         393.0   
43423  185933  

Removing rows having NaN in from_id

In [20]:
mod_data = mod_data[np.isfinite(mod_data['from_area_id'])]

In [21]:
mod_data

id  user_id  vehicle_model_id  travel_type_id  from_area_id  \
0      132512    22177                28               2          83.0   
1      132513    21413                12               2        1010.0   
2      132514    22178                12               2        1301.0   
3      132515    13034                12               2         768.0   
4      132517    22180                12               2        1365.0   
5      132518    17712                12               2        1021.0   
6      132519    22172                12               1         571.0   
7      132520    22181                12               2        1192.0   
8      132521    22182                65               3         448.0   
9      132522    22184                12               2         516.0   
10     132523     4941                12               2         150.0   
11     132524    17037                12               2         455.0   
12     132525    22185                12               2        1166.0   
13     132526    22186                12               2         793.0   
14     132527       16                28               2        1063.0   
15     132528    15097                12               2        1102.0   
16     132530      761                12               2         814.0   
17     132531    22189                24               1        1383.0   
18     132533    17226                12               2        1353.0   
19     132534      868                12               2         297.0   
20     132535    22190                87               3         471.0   
21     132536       16                28               2          58.0   
22     132537    21558                12               2         540.0   
23     132538    21995                12               3        1286.0   
24     132539    22192                54               1         515.0   
25     132540    22193                12               2         142.0   
26     132541    20430                12               2         393.0   
27     132542    22194                12               2          89.0   
28     132543    21716                28               2        1237.0   
29     132544    38966                12               2          61.0   
...       ...      ...               ...             ...           ...   
43401  185909    27899                65               3        1038.0   
43402  185910    47996                89               3         328.0   
43403  185911    41939                65               3        1038.0   
43404  185912    48711                12               2         238.0   
43405  185914    41939                12               3        1281.0   
43406  185915    40643                12               3         136.0   
43407  185916    48714                87               1        1180.0   
43408  185917    40643                12               3         136.0   
43409  185918    41939                12               3        1281.0   
43410  185919    41939                12               3        1281.0   
43411  185920    31846                65               3         541.0   
43412  185921    44027                12               2        1384.0   
43413  185922    48718                12               2        1377.0   
43414  185923    34127                12               2         376.0   
43415  185924    48719                12               3         768.0   
43416  185926    36929                12               2         396.0   
43417  185927    41939                12               3        1390.0   
43418  185928    48721                12               2        1263.0   
43419  185929    46807                12               2         571.0   
43420  185930    36265                65               3        1237.0   
43421  185931    48723                12               2        1371.0   
43422  185932    48724                12               2         393.0   
43423  185933  

In [30]:
mod_data['from_city_id'].unique()

array([nan, 15.,  1., 31.])

In [26]:
(mod_data['from_city_id']==15).sum()

16146

In [27]:
(mod_data['from_city_id']==1).sum()

106

In [28]:
(mod_data['from_city_id']==31).sum()

5

It is seen that maximum rides are from city id 15 and the huge number of NaN values indicate that city id wasn't recorded because it was evident. Drop from_city_id. Also, large number of NaN values in to_city_id eliminates the use of it.

In [32]:
mod_data=data.drop(['from_city_id'],axis=1)

In [35]:
mod_data=data.drop(['to_city_id'],axis=1)

In [41]:
data['from_area_id'].unique().shape

(599,)

In [40]:
data['from_area_id'].value_counts()

393.0     3858
571.0     1631
293.0     1052
585.0      911
1010.0     768
142.0      727
83.0       719
1384.0     628
1096.0     542
58.0       466
269.0      413
1026.0     395
625.0      370
1017.0     363
1330.0     362
689.0      341
768.0      334
271.0      333
1056.0     330
1371.0     327
1068.0     324
363.0      308
396.0      289
1021.0     279
1365.0     273
105.0      271
410.0      266
458.0      261
515.0      251
776.0      249
          ... 
1209.0       1
1277.0       1
1042.0       1
178.0        1
1158.0       1
1343.0       1
1164.0       1
483.0        1
802.0        1
1113.0       1
1215.0       1
1127.0       1
72.0         1
1253.0       1
384.0        1
770.0        1
1092.0       1
185.0        1
79.0         1
215.0        1
1235.0       1
68.0         1
138.0        1
369.0        1
1252.0       1
1178.0       1
1265.0       1
1345.0       1
1094.0       1
133.0        1
Name: from_area_id, Length: 598, dtype: int64

In [44]:
row=data['from_area_id']==393.0

In [47]:
data[row]

id  user_id  vehicle_model_id  package_id  travel_type_id  \
26     132541    20430                12         NaN               2   
48     132570    22211                65         NaN               2   
49     132571    22212                12         NaN               2   
57     132580    19469                85         NaN               2   
73     132600    21235                12         NaN               2   
90     132620    22236                85         2.0               3   
94     132625    18521                12         NaN               2   
96     132627    22239                85         NaN               2   
117    132650    22252                12         NaN               2   
125    132661    33542                12         NaN               2   
142    132678    18150                12         NaN               2   
143    132679    22267                12         NaN               2   
178    132718    18521                12         NaN               2   
201    132743    22301                12         NaN               2   
228    132773    18046                12         NaN               2   
232    132778    21102                12         NaN               2   
236    132783    40040                85         NaN               2   
238    132785    22321                12         NaN               2   
288    132851    22360                12         NaN               2   
293    132860     4778                12         NaN               2   
302    132872    17528                85         2.0               3   
310    132886     5871                12         NaN               2   
324    132903    18898                85         NaN               2   
328    132907    22385                12         NaN               1   
335    132914    22388                12         NaN               2   
339    132920    22390                12         NaN               2   
353    132938    22397                12         NaN               2   
367    132956    22406                23         NaN               2   
373    132963    41298                65         NaN               2   
382    132974     9483                12         NaN               2   
...       ...      ...               ...         ...             ...   
43110  185567    48524                23         NaN               2   
43120  185578    48530                12         NaN               2   
43121  185579    23869                90         4.0               3   
43152  185613    48550                12         NaN               2   
43156  185618    48554                12         NaN               2   
43165  185633    48562                65         NaN               1   
43174  185644    45712                12         NaN               2   
43175  185645    48571                12         NaN               2   
43185  185659    39513                12         NaN               2   
43195  185670    45903                12         NaN               2   
43204  185681    41503                12         NaN               2   
43209  185687    48593                65         NaN               2   
43212  185691    45604                12         NaN               2   
43250  185732    48614                87         NaN               1   
43254  185736    45712                12         NaN               2   
43260  185745    48621                28         NaN               2   
43278  185768    48631                12         NaN               2   
43281  185773    48248                12         NaN               2   
43288  185782    48640                12         NaN               2   
43289  185783    48641                12         NaN               2   
43292  185786    48363                24         NaN               2   
43296  185791    41006                12         NaN               2   
43325  185825    41488                12         NaN               2   
43342  185843    48667                12         Na

In [49]:
data['package_id'].value_counts()

1.0    3503
2.0    2651
6.0     502
4.0     412
3.0     375
7.0     101
5.0       6
Name: package_id, dtype: int64